Include cluster dropping functionality

jwaiton opened the pull request:
This PR resolves the issue with clusters of spurious hits occurring within the data by introducing a function (`drop_isolated_clusters()`) that replaces the previous `drop_isolated_sensors()`.

This function first finds all hits with more than `cluster_size` neighbours within the `drop_dist` distances in [X.,Y.,Z.]. It then iterates over these hits, ensuring that their neighbours are also retained (avoiding the loss of edge hits with fewer neighbours than the `cluster_size`).

This addition does not break backwards compatibility, as the `drop_isolated_sensors()` function is still used if the `drop_dist` provided is 2 dimensional, rather than 3 dimensional. This is a lousy solution to the problem and an alternative would be appreciated.

A simple test to check the functionality has been written and runs as expected, but I believe more rigorous testing is required.

jwaiton committed:
include drop_isolated_clusters

jwaiton committed:
implement into beersheba

jwaiton committed:
add test

jwaiton committed:
add `TypeError` and refactor arguments

jwaiton committed:
implement faster method using graphs

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -336,20 +337,31 @@ def cut_over_Q(df):  # df shall be an event cdst
-def drop_isolated(distance, redist_var):
+def drop_isolated( distance   : List[float],
+                   redist_var : Optional[List] = [],
+                   nhits      : Optional[int] = 3):
```
I would rather keep `redist_var` with no default value. Probably also give `nhits` a default of `None` and check its value if 3d is requested.

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -503,7 +515,7 @@ def beersheba( files_in         : OneOrManyFiles
     cut_sensors           = fl.map(cut_over_Q   (deconv_params.pop("q_cut")    , ['E', 'Ec']),
                                    item = 'hits')
-    drop_sensors          = fl.map(drop_isolated(deconv_params.pop("drop_dist"), ['E', 'Ec']),
+    drop_sensors          = fl.map(drop_isolated(deconv_params.pop("drop_dist"), ['E', 'Ec'], deconv_params.pop("cluster_size")),
```
I guess you need to add "cluster_size" to `deconv_params` if not given, right?

gonzaponte replied:
Also, do we need to include this in esmeralda as well?

jwaiton replied:
Esmeralda doesn't have `drop_isolated()` as a function. I included it within the `thekla` city for classical analysis purposes to avoid altering `esmeralda` when others are using it.

jwaiton replied:
> I guess you need to add "cluster_size" to `deconv_params` if not given, right?

Yes, good point! I had a couple of ideas to alter this for backwards compatibility purposes. I'm currently looking at this as a possibility.

```
drop_sensors = fl.map(drop_isolated(deconv_params.pop("drop_dist"), ['E', 'Ec'], deconv_params.pop("cluster_size") if "cluster_size" in deconv_params else None),
                      item = 'hits')
```

But I'm open to better suggestions.

gonzaponte replied:
Look at the docs for `dict.pop`, you'll find that it accepts a second argument which is the default value in case the key is not present in the dictionary.

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -181,6 +183,66 @@ def drop_isolated_sensors(df : pd.DataFrame) -> pd.DataFrame:
 
+def drop_isolated_clusters(distance   :  List[float]= [10., 10., 1.],
+                           nhits      :  int        = 3,
+                           variables  :  List[str  ]= [      ]) -> Callable:
```
No default values, please.

gonzaponte replied:
This follows more the components style (a function that fixes the configuration for some other inner function) and therefore should go in `components.py`. I've seen that `drop_isolated_sensors` does the same, so perhaps we should do the same with that one. But that's for later. Now, here I think it would be desirable to have the inner function take all 4 arguments in the file scope so it can be tested in a more direct manner and the outer one would only need a simple test.

jwaiton replied:
`drop_isolated_sensors()` and `cut_and_redistribute_df()` also follow this method of having arguments in the outer function as 'initialisation parameters'. Should I modify these as well? Or perhaps leave that for another PR. 

Putting all 4 arguments into the inner argument adds a bit of complexity to how the function is then applied to the data. For example, in this case the [line](https://github.com/next-exp/IC/blob/master/invisible_cities/cities/beersheba.py#L354):
`df = df.groupby(['event', 'npeak']).apply(drop).reset_index(drop=True)`,

would (I think) require modification to:
`df = df.groupby(['event', 'npeak']).apply(lamda x: drop(x, distance, nhits, redist_var, arg2, arg3)).reset_index(drop=True)`.

In which case, why not just remove the outer function? If it doesnt 'initalise' the function for use, it serves no purpose as far as I can tell (in this case).

gonzaponte replied:
> drop_isolated_sensors() and cut_and_redistribute_df() also follow this method of having arguments in the outer function as 'initialisation parameters'. Should I modify these as well? Or perhaps leave that for another PR.

Let's leave all that for another PR

> Putting all 4 arguments into the inner argument adds a bit of complexity to how the function is then applied to the data.

I prefer keeping the current structure

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -181,6 +183,66 @@ def drop_isolated_sensors(df : pd.DataFrame) -> pd.DataFrame:
+        x   = df.X.values / distance[0]
+        y   = df.Y.values / distance[1]
+        z   = df.Z.values / distance[2]
+        xyz = np.column_stack((x,y,z))
```
you can take advantage of numpy's broadcasting rules
```suggestion
        # distance should be a np.array before this operation
        xyz = df[list("XYZ")].values / distance # or df[["X", "Y", "Z"]] or df["X Y Z".split()]
```

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -181,6 +183,66 @@ def drop_isolated_sensors(df : pd.DataFrame) -> pd.DataFrame:
+        # create graph that connects all close pairs between hit positions based on df index
+        cluster_graph = nx.Graph()
+        cluster_graph.add_nodes_from(range(len(df)))
+        cluster_graph.add_edges_from((df.index[i], df.index[j]) for i,j in pairs)
```
from the comment above the definition of `pairs` I undertand that `pairs` contains less elements than `df`. Is that the case? If so, is it ok to have more nodes than edges? If not, maybe clarify the comments.

jwaiton replied:
You're correct `df` should generally have more entries than the number of pairs, and as far as I understand there is no issue with having more nodes than edges.

The `pair` indexing (`i,j`) only relates to the relevant nodes they are taken from within the `cKDTree`, as far as I understand (it's been a while, but I believe I tested this to ensure it was correct). So the first case of `i,j` (if unsorted) may be `2, 45`, relating to the distance between `df[2]` and `df[45]` being smaller than `distance`.

Should I write a better explanation of how it works in the comments/docstring? It's currently quite barebones I must admit.

gonzaponte replied:
I think I understand, isolated nodes are there, but since they do not have any link to other nodes they are effectively ignored by the algorithm. I think it's clear, not sure what I was missing when I left this comment.

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -181,6 +183,66 @@ def drop_isolated_sensors(df : pd.DataFrame) -> pd.DataFrame:
+        cluster_graph.add_edges_from((df.index[i], df.index[j]) for i,j in pairs)
+
+        # Find all clusters within the graph
+        clusters = list(nx.connected_components(cluster_graph))
```
no need to consume the iterator, is there?

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -181,6 +183,66 @@ def drop_isolated_sensors(df : pd.DataFrame) -> pd.DataFrame:
+        clstrs = []
+        for cluster in clusters:
+            if len(cluster) > nhits:
+                passing_hits |= cluster
```
I might be (heavily) biased, but how about this? Is it difficult to parse?
```suggestion
        passing_hits = reduce(set.union, filter(lambda x: len(x)>nhits, clusters))
```

jwaiton replied:
Its a bit harder to read I'd argue, but only due to a lack of knowledge on my part (I didn't know exactly how `reduce` and `filter` works before checking).

Regardless, I like your method more (I think its useful to learn).

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -181,6 +183,66 @@ def drop_isolated_sensors(df : pd.DataFrame) -> pd.DataFrame:
+ 
+        # extract mask and apply it
+        mask    = df.index.isin(passing_hits)
+        pass_df = df.loc[mask, :].copy()
```
If I understand correctly what's happening here, this
```suggestion
        pass_df = df.loc[passing_hits, :].copy()
```
should be the same

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -103,6 +104,32 @@ def test_drop_isolated_sensors():
+    # check the shape is as expected, E isn't the same
+    assert df_cut[['X', 'Y', 'Z', 'Q']].equals(df.head(5)[['X', 'Y', 'Z', 'Q']])
+
+
```
- you should split this test into two: one checks energy conservation, the other number of neighbors. Use a fixture if you don't want to repeat the dataframe (it's only twice, so it's not too bad to repeat it either)
- don't use random numbers, it breaks reproducibility
- why `if len(df_cut)`? isn't the output perfectly predictable?
- the last assertion is a bit tedious to read. you could split it into three lines like
```
got      = ...
expected = ...
assert got.equals(expected)
```

jwaiton replied:
This test is more or less copied from `test_drop_isolated_sensors()`. The `if len(df_cut)` component is extracted from that test, which I think it is required due to it using random assignment for the hit positions. Although I'm not sure if it makes sense in either test. 

Should I alter `test_drop_isolated_sensors()` to remove the randomness as well? The test itself is a bit questionable, it relies on randomness that I assume would cause it to occasionally fail but I haven't seen it fail in my testing.

gonzaponte replied:
Focus on this one first. I would like to remove randomness from tests overall, but this is orthogonal to this PR. If you fancy doing it for `drop_isolated_sensors` as well, go ahead, but in a separate commit.

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -181,6 +183,66 @@ def drop_isolated_sensors(df : pd.DataFrame) -> pd.DataFrame:
+
+        # build KDTree of datapoints, collect pairs within distance
+        xyz_tree = cKDTree(xyz)
+        pairs    = xyz_tree.query_pairs(r=dist)
```
How about we compact this a bit?
```suggestion
        # build KDTree of datapoints, collect pairs within distance
        pairs    = cKDTree(xyz).query_pairs(r= np.sqrt(3)) # meaning of sqrt(3)
```

gonzaponte commented on invisible_cities/config/beersheba.conf:
```diff
@@ -17,6 +17,7 @@ same_peak = True
 deconv_params = dict(
   q_cut         = 10,
   drop_dist     = [10., 10.],
+  cluster_size  = 1,
```
if 2d, this parameter should not be given... Unless the 2d case can be generalized to work with different cluster sizes.

gonzaponte commented on invisible_cities/conftest.py:
```diff
@@ -840,12 +840,13 @@ def beersheba_config(Th228_hits, PSFDIR, next100_mc_krmap):
                  , same_peak     = True
                  , deconv_params = dict( q_cut         = 10
-                                       , drop_dist     = [16.0] * 2
+                                       , drop_dist     = [16.0, 16.0, 4.]
                                        , psf_fname     = PSFDIR
                                        , e_cut         = 12e-3
                                        , n_iterations  = 100
                                        , iteration_tol = 1e-10
                                        , sample_width  = [15.55] * 2
+                                       , cluster_size  = 3
                                        , bin_size      = [ 1.,  1.]
                                        , diffusion     = (1.0, 0.2)
```
This means that the tests now run `drop_isolated_clusters` instead of `drop_isolated_sensors`, right? I'm amazed this didn't break the tests. I think we should keep doing what we did before, no? And eventually add a beersheba test with 3d cluster dropping.

gonzaponte reviewed (changes requested):
I had a first go at this. There is the question of whether we can reproduce the behaviour of `drop_isolated_sensors` using `drop_isolated_clusters` in 2d. If so, we should try to merge them, although not necessarily in this PR.

jwaiton committed:
add catch for missing `cluster_size`

This should allow for backwards compatibility (older configs without `cluster_size` should work)

jwaiton committed:
remove default arguments for `drop_isolated_clusters`

jwaiton committed:
refactor improving readibility, removing redundant unpacking and variable assignment.

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -336,20 +337,34 @@ def cut_over_Q(df):  # df shall be an event cdst
+    distance   : Sensor pitch in 2 or 3 dimensions
     redist_var : List with variables to be redistributed.
-
+    nhits      : Number of hits 
```
This requires a bit more detail to be understood

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -336,20 +337,34 @@ def cut_over_Q(df):  # df shall be an event cdst
     Parameters
     ----------
-    distance   : Sensor pitch.
+    distance   : Sensor pitch in 2 or 3 dimensions
```
Sensor pitch in 3d does not make sense, Perhaps "distance between hits"?

gonzaponte reviewed (commented):
a couple of minor doc comments

jwaiton committed:
update docstrings and cosmetics

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -181,6 +185,54 @@ def drop_isolated_sensors(df : pd.DataFrame) -> pd.DataFrame:
+
+    Parameters
+    ----------
+    df       : Groupby ('event' and 'npeak') dataframe
```
```suggestion
    df : Groupby ('event' and 'npeak') dataframe
```

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -181,6 +185,54 @@ def drop_isolated_sensors(df : pd.DataFrame) -> pd.DataFrame:
+                           nhits      :  int,
+                           variables  :  List[str  ]) -> Callable:
+    '''
+    Drops isolated clusters of hits (SiPMs).
```
Document a bit the algorithm in general terms, not only for you, but for future users.

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -336,20 +337,35 @@ def cut_over_Q(df):  # df shall be an event cdst
-    Drops rogue/isolated hits (SiPMs) from hits.
+    Drops rogue/isolated hits (SiPMs) from hits,
+    can be configured to remove isolated clusters
+    below a certain threshold number of hits. 
```
To keep the style a bit more homogeneous try to use a wider textwidth, up to 70 or 80 characters per line (counting indentation)

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -103,6 +104,52 @@ def test_drop_isolated_sensors():
+    df_cut        = drop_function(df)
+
+    got           = df_cut[['X', 'Y', 'Z', 'Q']]
+    expected      = df.head(5)[['X', 'Y', 'Z', 'Q']]
```
```suggestion
    got           = df_cut    [['X', 'Y', 'Z', 'Q']]
    expected      = df.head(5)[['X', 'Y', 'Z', 'Q']]
```

gonzaponte reviewed (commented):
I'm very happy with this, I missed a couple of minor style and doc issues last time. After that, I'll approve the PR.

gonzaponte reviewed (approved):
This PR adds a new function to drop clusters, which is very similar to the existing `drop_isolated_sensors`. Both functions have been combined into an interfacing function. The new function is tested.

Good job!
